pandas/core/interval.py

```diff
@@ -0,0 +1,413 @@
+
+    def _get_loc_only_exact_matches(self, key):
+        return self._multiindex._tuple_index.get_loc(key)
+
+    def _find_non_overlapping_monotonic_bounds(self, key):
+        if isinstance(key, IntervalMixin):
+            start = self._searchsorted_monotonic(
+                key.left, 'left', exclude_label=key.open_left)
+            stop = self._searchsorted_monotonic(
+                key.right, 'right', exclude_label=key.open_right)
+        else:
+            # scalar
+            start = self._searchsorted_monotonic(key, 'left')
+            stop = self._searchsorted_monotonic(key, 'right')
+        return start, stop
+
+    def get_loc(self, key, method=None):
+        self._check_method(method)
+
+        original_key = key
+
+        if self.is_non_overlapping_monotonic:
+            if isinstance(key, Interval):
+                left = self._maybe_cast_slice_bound(key.left, 'left', None)
+                right = self._maybe_cast_slice_bound(key.right, 'right', None)
+                key = Interval(left, right, key.closed)
+            else:
+                key = self._maybe_cast_slice_bound(key, 'left', None)
+
+            start, stop = self._find_non_overlapping_monotonic_bounds(key)
+
+            if start + 1 == stop:
+                return start
+            elif start < stop:
+                return slice(start, stop)
+            else:
+                raise KeyError(original_key)
+
+        else:
+            # use the interval tree
+            if isinstance(key, Interval):
+                left, right = _get_interval_closed_bounds(key)
+                return self._tree.get_loc_interval(left, right)
+            else:
+                return self._tree.get_loc(key)
+
+    def get_value(self, series, key):
+        # this method seems necessary for Series.__getitem__ but I have no idea
+        # what it should actually do here...
+        loc = self.get_loc(key)  # nb. this can't handle slice objects
+        return series.iloc[loc]
+
+    def get_indexer(self, target, method=None, limit=None):
+        self._check_method(method)
+        target = _ensure_index(target)
+
+        if self.is_non_overlapping_monotonic:
+            start, stop = self._find_non_overlapping_monotonic_bounds(target)
+
+            start_plus_one = start + 1
+            if (start_plus_one < stop).any():
+                raise ValueError('indexer corresponds to non-unique elements')
+            return np.where(start_plus_one == stop, start, -1)
+
+        else:
+            if isinstance(target, IntervalIndex):
+                raise NotImplementedError(
+                    'have not yet implemented get_indexer '
+                    'for IntervalIndex indexers')
+            else:
+                return self._tree.get_indexer(target)
+
+    def delete(self, loc):
+        new_left = self.left.delete(loc)
+        new_right = self.right.delete(loc)
+        return type(self)(new_left, new_right, self.closed, self.name,
```

Comment:
Note that `_shallow_copy` is way slower for `IntervalIndex`, because it needs to use `.values`, which is the 1d object array. That's why I've rewritten these methods to use `.left` and `.right` instead.